packages/dds/tree/src/feature-libraries/sequence-field/tracker.ts

```diff
@@ -0,0 +1,102 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/common-utils";
+import { RevisionTag } from "../../core";
+import { Mark } from "./format";
+import {
+    getInputLength,
+    getOutputLength,
+    isAttach,
+    isDetachMark,
+    isNetZeroNodeCountChange,
+} from "./utils";
+
+export class IndexTracker {
+    private inputIndex: number = 0;
+    private readonly contributions: { rev: RevisionTag; nodes: number }[] = [];
+
+    public advance(mark: Mark<unknown>): void {
+        const inLength = getInputLength(mark);
+        const outLength = getOutputLength(mark);
+        this.inputIndex += inLength;
+        if (isNetZeroNodeCountChange(mark)) {
+            return;
+        }
+        const nodes = outLength - inLength;
+        // If you hit this assert, then you probably need to add a check for it in `isNetZeroNodeCountChange`.
+        assert(nodes !== 0, "Unknown mark type with net-zero node count change");
+        const revision = mark.revision;
+        // TODO: Remove this early return. It is only needed because some tests use anonymous changes.
+        // These tests will fail (i.e., produce the wrong result) if they rely the index tracking performed here.
+        if (revision === undefined) {
+            return;
+        }
+        assert(revision !== undefined, "Compose base mark should carry revision info");
+        const index = this.contributions.findIndex(({ rev }) => rev > revision);
+        if (index === -1) {
+            this.contributions.push({ rev: revision, nodes });
+        } else {
+            if (this.contributions[index].rev !== revision) {
```

Comment:
Indeed. That was a mistake. Interestingly that mistake only had a perf impact: it added extra entries when an existing entry could be used instead. This had not correctness impact because the `get` algorithm for this class would either stop before the first such entry iterate over them all, summing as it goes.